src/lib/scheduler/operator_task.cpp

```diff
@@ -30,20 +30,27 @@ std::vector<std::shared_ptr<AbstractTask>> OperatorTask::make_tasks_from_operato
 std::shared_ptr<AbstractTask> OperatorTask::_add_tasks_from_operator(
     const std::shared_ptr<AbstractOperator>& op, std::vector<std::shared_ptr<AbstractTask>>& tasks,
     std::unordered_map<std::shared_ptr<AbstractOperator>, std::shared_ptr<AbstractTask>>& task_by_op) {
+  // Early out: We do not want to re-run operators, if they have been executed by previous tasks.
+  if (op->performance_data->executed) return nullptr;
+
   const auto task_by_op_it = task_by_op.find(op);
   if (task_by_op_it != task_by_op.end()) return task_by_op_it->second;
 
   auto task = std::make_shared<OperatorTask>(op);
   task_by_op.emplace(op, task);
 
   if (auto left = op->mutable_left_input()) {
-    auto subtree_root = _add_tasks_from_operator(left, tasks, task_by_op);
-    subtree_root->set_as_predecessor_of(task);
+    if (!left->performance_data->executed) {
```

Comment:
Wouldn't this be caught once we recurse into `_add_tasks_from_operator`? I wouldn't add the early out twice.